File: kotlin/basics/07_async.kt
package basics
// TOPIC: Coroutines & Async | kotlinc 07_async.kt -include-runtime -d out.jar && java -jar out.jar
// Requires: kotlinx-coroutines-core — add to classpath or use Gradle/Maven
// Docs: https://kotlinlang.org/docs/coroutines-overview.html

fun main() {
    // TODO 1: runBlocking + launch — basic coroutine, delay instead of sleep
    //   - runBlocking { } to bridge blocking world and coroutines
    //   - launch { delay(1000); println("done") } — fire-and-forget
    //   - delay() is non-blocking (suspends coroutine, frees thread)
    //   - join() a Job to wait for it: val job = launch { ... }; job.join()

    // TODO 2: async/await — async { fetchUser() }, async { fetchPosts() }, awaitAll()
    //   - val userDeferred = async { fetchUser() }
    //   - val postsDeferred = async { fetchPosts() }
    //   - Both run concurrently; await each: userDeferred.await()
    //   - awaitAll(userDeferred, postsDeferred) — wait for all at once
    //   - Measure total time: should be max(t1, t2) not t1+t2

    // TODO 3: Flows — flow { emit(1); emit(2) }, collect, map, filter, flowOn
    //   - val numberFlow = flow { for (i in 1..5) { emit(i); delay(100) } }
    //   - numberFlow.collect { println(it) }
    //   - numberFlow.map { it * 2 }.filter { it > 4 }.collect { ... }
    //   - flowOn(Dispatchers.IO) to change upstream context
    //   - Terminal operators: collect, toList, first, single, reduce

    // TODO 4: Channels — producer-consumer with Channel<Int>(capacity=10)
    //   - val channel = Channel<Int>(capacity = 10)
    //   - Producer coroutine: channel.send(item); channel.close()
    //   - Consumer coroutine: for (item in channel) { process(item) }
    //   - produce { } builder creates a ReceiveChannel
    //   - Channel types: UNLIMITED, BUFFERED, RENDEZVOUS, CONFLATED

    // TODO 5: Structured concurrency — coroutineScope, supervisorScope, cancellation propagation
    //   - coroutineScope { }: fails fast — any child failure cancels all siblings
    //   - supervisorScope { }: isolated failures — siblings continue on child failure
    //   - Cancellation is cooperative: check isActive or use cancellable suspend fns
    //   - withTimeout(3000L) { ... } — cancels block after timeout

    // TODO 6: withContext — switch dispatchers: IO, Default, Main
    //   - Dispatchers.IO: for blocking I/O (file, network, DB)
    //   - Dispatchers.Default: for CPU-intensive work (sorting, parsing)
    //   - Dispatchers.Main: for UI updates (Android)
    //   - withContext(Dispatchers.IO) { readFile() } — suspends, doesn't block

    // TODO 7: StateFlow & SharedFlow — hot flows for state management
    //   - val stateFlow = MutableStateFlow(0) — always has current value, replays to new collectors
    //   - stateFlow.value = 1; stateFlow.update { it + 1 }
    //   - val sharedFlow = MutableSharedFlow<Event>() — no initial value, configurable replay
    //   - sharedFlow.emit(Event.Click); sharedFlow.collect { handle(it) }
    //   - Use case: StateFlow for UI state, SharedFlow for one-time events

    // TODO 8: Mutex — Mutex().withLock { } for coroutine-safe shared state
    //   - val mutex = Mutex()
    //   - var counter = 0
    //   - launch { mutex.withLock { counter++ } }
    //   - Compare with atomic operations: AtomicInteger (but prefer structured concurrency)
    //   - Show race condition without mutex, then fix it
}

File: ruby/basics/07_async.rb
# TOPIC: Threads & Fibers | ruby 07_async.rb
# Docs: https://ruby-doc.org/core/Thread.html

# TODO 1: Thread.new — launch threads, join, capture return value
#   threads = (1..5).map { |i| Thread.new { sleep(rand * 0.1); i * i } }
#   results = threads.map(&:value)
#   puts results.inspect
#   - Thread#value blocks until thread finishes and returns the block result

# TODO 2: Mutex for thread-safe shared state
#   mutex = Mutex.new
#   counter = 0
#   threads = 10.times.map { Thread.new { 1000.times { mutex.synchronize { counter += 1 } } } }
#   threads.each(&:join)
#   puts counter  # should be 10_000 (not less due to race conditions)

# TODO 3: Queue (thread-safe) for producer-consumer
#   require 'thread'
#   queue = Queue.new
#   producer = Thread.new { 10.times { |i| queue.push(i); sleep 0.01 } }
#   consumer = Thread.new { loop { item = queue.pop; break if item == :stop; puts "Got #{item}" } }
#   producer.join; queue.push(:stop); consumer.join

# TODO 4: Fiber — cooperative coroutines (manual yield/resume)
#   fib = Fiber.new do
#     a, b = 0, 1
#     loop { Fiber.yield(a); a, b = b, a + b }
#   end
#   10.times { print "#{fib.resume} " }
#   - Fibers are NOT threads — single-threaded, explicit scheduling

# TODO 5: Ractor (Ruby 3+) — parallel execution without GVL
#   r = Ractor.new { Ractor.recv * 2 }
#   r.send(21)
#   puts r.take  # => 42
#   - Only Ractors can share state safely; objects must be shareable

# TODO 6: Thread.abort_on_exception
#   Thread.abort_on_exception = true  # unhandled exception in thread kills main program
#   - Default: exceptions in threads are silently swallowed unless joined
#   - Alternative: Thread.report_on_exception = true (Ruby 2.4+)

# TODO 7: Async gem — non-blocking I/O with fibers (external gem)
#   # require 'async'
#   # Async do |task|
#   #   5.times { task.async { |t| t.sleep(rand); puts "done" } }
#   # end

# TODO 8: GVL (Global VM Lock) explanation
#   # MRI Ruby has a GVL — only one thread runs Ruby code at a time
#   # Threads CAN run in parallel during C extensions and I/O (e.g., sleep, HTTP)
#   # For CPU-parallel work: use Ractor (Ruby 3+) or fork processes
#   # Demonstrate: two CPU-bound threads vs two I/O-bound threads
